File: extract.js
const fs = require('fs');
const lines = fs.readFileSync('C:/Users/AD/.gemini/antigravity/brain/f388562f-c6f2-493b-80a2-77824d1f6624/.system_generated/logs/transcript_full.jsonl', 'utf8').trim().split('\n');
let found = false;
for (let i = lines.length - 1; i >= 0; i--) {
  const line = JSON.parse(lines[i]);
  if (line.content && line.content.includes('"sources":')) {
    const start1 = line.content.indexOf('{\n  "sources":');
    const start2 = line.content.indexOf('{"sources":');
    const start = Math.max(start1, start2);
    const end = line.content.lastIndexOf('}');
    if (start !== -1) {
      const jsonStr = line.content.substring(start, end + 1);
      fs.writeFileSync('frontend/src/data/bai_tap.json', jsonStr);
      console.log('Extracted! Length:', jsonStr.length);
      found = true;
      break;
    } else {
      console.log('Found string but no block', line.content.substring(0, 100));
    }
  }
}
if (!found) {
  // try regex
  for (let i = lines.length - 1; i >= 0; i--) {
    const line = JSON.parse(lines[i]);
    if (line.content) {
      const match = line.content.match(/\{[\s\S]*?"sources"[\s\S]*?\}/);
      if (match) {
        // extract till the LAST }
        const start = match.index;
        const end = line.content.lastIndexOf('}');
        const jsonStr = line.content.substring(start, end + 1);
        fs.writeFileSync('frontend/src/data/bai_tap.json', jsonStr);
        console.log('Extracted via regex! Length:', jsonStr.length);
        found = true;
        break;
      }
    }
  }
}

File: extract.py
import json
import re

with open(r'C:\Users\AD\.gemini\antigravity\brain\f388562f-c6f2-493b-80a2-77824d1f6624\.system_generated\logs\transcript_full.jsonl', 'r', encoding='utf-8') as f:
    lines = f.readlines()

for line in reversed(lines):
    data = json.loads(line)
    if data.get('type') == 'USER_INPUT':
        content = data.get('content', '')
        if '"sources": [' in content:
            # find the start of json
            start_idx = content.find('{\n  "sources":')
            if start_idx == -1:
                start_idx = content.find('{"sources":')
            end_idx = content.rfind('}')
            if start_idx != -1 and end_idx != -1:
                json_str = content[start_idx:end_idx+1]
                with open('frontend/src/data/bai_tap.json', 'w', encoding='utf-8') as out:
                    out.write(json_str)
                print(f"Extracted json of length {len(json_str)}")
                break
